crates/wasmparser/src/validator/operators.rs

```diff
@@ -2156,122 +2167,107 @@ impl OperatorValidator {
-                            if !self.control.last().unwrap().unreachable {
-                                bail_op_err!(
-                                    "type mismatch: call_ref only works on index-type references"
-                                )
-                            }
+                        for ty in ft.outputs() {
+                            self.push_operand(ty, resources)?;
                         }
-                    }
+                    },
+                    HeapType::Bot => (),
+                    _ => bail_op_err!("type mismatch: instruction requires function reference type but stack has {}", ty_to_str(ValType::Ref(rt)))
                 }
             }
             Operator::ReturnCallRef => {
                 self.check_function_references_enabled()?;
-                if let Some(rt) = self.pop_ref(resources)? {
-                    match rt.heap_type {
-                        HeapType::Index(type_index) => {
-                            let ty = func_type_at(resources, type_index)?;
-                            for ty in ty.inputs().rev() {
-                                self.pop_operand(Some(ty), resources)?;
-                            }
-                            for ty in ty.outputs() {
-                                self.push_operand(ty, resources)?;
-                            }
+                // if !self.features.tail_call {
+                //     return Err(OperatorValidatorError::new(
+                //         "tail calls support is not enabled",
+                //     ));
+                // }
+                let rt = self.pop_ref(resources)?;
+                match rt.heap_type {
+                    HeapType::Index(type_index) => {
+                        let ft = func_type_at(resources, type_index)?;
+                        for ty in ft.inputs().rev() {
+                            self.pop_operand(Some(ty), resources)?;
                         }
-                        _ => {
-                            if !self.control.last().unwrap().unreachable {
-                                bail_op_err!(
-                                    "type mismatch: call_ref only works on index-type references"
-                                )
-                            }
+                        for ty in ft.outputs() {
+                            self.push_operand(ty, resources)?;
                         }
-                    }
+                    },
+                    HeapType::Bot => (),
+                    _ => bail_op_err!("type mismatch: instruction requires function reference type but stack has {}", ty_to_str(ValType::Ref(rt)))
                 }
                 self.check_return(resources)?;
             }
             Operator::RefAsNonNull => {
                 self.check_function_references_enabled()?;
-                if let Some(RefType { heap_type, .. }) = self.pop_ref(resources)? {
-                    self.check_heap_type(heap_type, resources)?;
```

Comment:
Ah, I see, push_operand should take care of this

Reply:
Yep! The validation is much more uniform/simpler since `Bot` is part of the type structure, and not external as it were with the `Option` representation.